compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/ExpectedTypeFinder.java

```diff
@@ -0,0 +1,1138 @@
+                    return getExpectedType(argument.get());
+                }
+            }
+
+            return getExpectedType(bLangNode);
+        }
+
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(ListConstructorExpressionNode node) {
+        BLangNode bLangNode = nodeFinder.lookup(this.bLangCompilationUnit, node.lineRange());
+        if (bLangNode == null) {
+            return Optional.empty();
+        }
+
+        Optional<TypeSymbol> extractedType = getExpectedType(bLangNode);
+        if (!extractedType.isPresent()) {
+            return Optional.empty();
+        }
+
+        if (extractedType.get().typeKind() != TypeDescKind.ARRAY) {
+            throw new IllegalStateException();
+        }
+
+        if (isWithinParenthesis(linePosition, node.openBracket(), node.closeBracket())) {
+            BArrayType bArrayType = (BArrayType) ((BallerinaArrayTypeSymbol) extractedType.get()).getBType();
+            if (bArrayType.eType != null && bArrayType.eType.getKind() != TypeKind.OTHER) {
+                return Optional.of(typesFactory.getTypeDescriptor(bArrayType.eType));
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(CaptureBindingPatternNode node) {
+        Optional<Symbol> variableSymbol = this.getSymbolByName(node.variableName().text());
+        if (variableSymbol.isEmpty()) {
+            return Optional.empty();
+        }
+        
+        BLangNode bLangNode = nodeFinder.lookup(this.bLangCompilationUnit, node.lineRange());
+        if (bLangNode == null) {
+            return Optional.empty();
+        }
+
+        if (bLangNode.getBType().getKind() == TypeKind.MAP) {
+            return  getTypeFromBType(bLangNode.getBType());
+        }
+
+        BType bType = bLangNode.getBType();
+        if (bLangNode.getBType().getKind() == TypeKind.OTHER) {
+            return Optional.empty();
+        }
+
+        return Optional.of(typesFactory.getTypeDescriptor(bType));
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(MappingConstructorExpressionNode node) {
+        return node.parent().apply(this);
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(WaitActionNode node) {
+        BLangNode bLangNode = nodeFinder.lookup(this.bLangCompilationUnit, node.lineRange());
+        if (bLangNode == null) {
+            return Optional.empty();
+        }
+
+        if (bLangNode instanceof BLangWaitExpr) {
+            BLangWaitExpr waitExpr = (BLangWaitExpr) bLangNode;
+            for (BLangExpression bLangExpression : waitExpr.exprList) {
+                if (isPosWithinRange(linePosition, bLangExpression.getPosition().lineRange())) {
+                    return getExpectedType(bLangExpression);
+                }
+            }
+
+            return getExpectedType(waitExpr);
+        }
+
+        return Optional.empty();
```

Comment:
Check if something like this which has an early return could be used to reduce the number of nested levels.

```suggestion
        if (!(bLangNode instanceof BLangWaitExpr)) {
           return Optional.empty();
        }
         BLangWaitExpr waitExpr = (BLangWaitExpr) bLangNode;
         for (BLangExpression bLangExpression : waitExpr.exprList) {
             if (isPosWithinRange(linePosition, bLangExpression.getPosition().lineRange())) {
                 return getExpectedType(bLangExpression);
             }
         }
         
         return getExpectedType(waitExpr);
```